File: src-swift/Sources/MurmurCore/ASR/VolcengineProtocol.swift
import Foundation

/// Volcengine BigASR V3 binary protocol constants and helpers.
/// See: https://www.volcengine.com/docs/6561/1354869
public enum VolcengineProtocol {
    public static let endpoint = "wss://openspeech.bytedance.com/api/v3/sauc/bigmodel"

    // Message types (high nibble of byte 1)
    static let msgFullClientRequest:  UInt8 = 0x01
    static let msgAudioOnlyRequest:   UInt8 = 0x02
    static let msgFullServerResponse: UInt8 = 0x09
    static let msgServerAck:          UInt8 = 0x0B
    static let msgServerError:        UInt8 = 0x0F

    // Sequence flags (low nibble of byte 1)
    static let flagPosSequence: UInt8 = 0x01
    static let flagNegSequence: UInt8 = 0x03

    // Serialization (high nibble of byte 2)
    static let serialJson:   UInt8 = 0x01

    // Compression (low nibble of byte 2)
    static let compressNone: UInt8 = 0x00
    static let compressGzip: UInt8 = 0x01

    // MARK: - Header building

    public static func buildHeader(
        msgType: UInt8, msgFlags: UInt8, serial: UInt8, compress: UInt8
    ) -> Data {
        var d = Data(count: 4)
        d[0] = (0x01 << 4) | 0x01   // version=1, headerSize=1
        d[1] = (msgType << 4) | msgFlags
        d[2] = (serial << 4) | compress
        d[3] = 0x00
        return d
    }

    // MARK: - Int32 big-endian helpers

    public static func int32ToData(_ value: Int32) -> Data {
        var v = value.bigEndian
        return Data(bytes: &v, count: 4)
    }

    public static func dataToInt32(_ data: Data, offset: Int = 0) -> Int32 {
        var v: Int32 = 0
        withUnsafeMutableBytes(of: &v) { dest in
            _ = data.copyBytes(to: dest, from: offset..<(offset + 4))
        }
        return Int32(bigEndian: v)
    }

    // MARK: - Packet building

    /// Build the full init request packet (with gzip-compressed JSON payload).
    public static func buildInitPacket(payload: [String: Any], sequence: Int32) throws -> Data {
        let jsonData = try JSONSerialization.data(withJSONObject: payload)
        let compressed = try GzipUtils.compress(jsonData)

        var packet = buildHeader(
            msgType: msgFullClientRequest,
            msgFlags: flagPosSequence,
            serial: serialJson,
            compress: compressGzip
        )
        packet.append(int32ToData(sequence))
        packet.append(int32ToData(Int32(compressed.count)))
        packet.append(compressed)
        return packet
    }

    /// Build an audio chunk packet.
    public static func buildAudioPacket(audio: Data, sequence: Int32, isLast: Bool) -> Data {
        let flag: UInt8 = isLast ? flagNegSequence : flagPosSequence
        let seqValue: Int32 = isLast ? -sequence : sequence

        var packet = buildHeader(
            msgType: msgAudioOnlyRequest,
            msgFlags: flag,
            serial: serialJson,
            compress: compressNone
        )
        packet.append(int32ToData(seqValue))
        packet.append(int32ToData(Int32(audio.count)))
        packet.append(audio)
        return packet
    }

    // MARK: - Response parsing

    public struct ParsedResponse {
        public enum Kind { case ack, result, error }
        public var kind: Kind
        public var sequence: Int32
        public var text: String?
        public var isFinal: Bool
        public var errorMessage: String?
    }

    public static func parseResponse(_ data: Data) -> ParsedResponse? {
        guard data.count >= 4 else { return nil }
        let msgType  = (data[1] >> 4) & 0x0F
        let msgFlags =  data[1] & 0x0F
        let compress =  data[2] & 0x0F

        switch msgType {
        case msgServerError:
            guard data.count >= 12 else { return nil }
            let msgSize = Int(dataToInt32(data, offset: 8))
            guard msgSize >= 0, data.count >= 12 + msgSize else { return nil }
            let raw = data.subdata(in: 12..<(12 + msgSize))
            let msg: String
            if compress == compressGzip, let dec = try? GzipUtils.decompress(raw) {
                msg = String(data: dec, encoding: .utf8) ?? ""
            } else {
                msg = String(data: raw, encoding: .utf8) ?? ""
            }
            return ParsedResponse(kind: .error, sequence: 0, isFinal: false, errorMessage: msg)

        case msgServerAck:
            guard data.count >= 8 else { return nil }
            let seq = dataToInt32(data, offset: 4)
            return ParsedResponse(kind: .ack, sequence: seq, isFinal: false)

        case msgFullServerResponse:
            guard data.count >= 12 else { return nil }
            let seq = dataToInt32(data, offset: 4)
            let payloadSize = Int(dataToInt32(data, offset: 8))
            guard payloadSize >= 0, data.count >= 12 + payloadSize else { return nil }
            let rawPayload = data.subdata(in: 12..<(12 + payloadSize))
            let payloadData: Data
            if compress == compressGzip {
                guard let dec = try? GzipUtils.decompress(rawPayload) else { return nil }
                payloadData = dec
            } else {
                payloadData = rawPayload
            }
            guard let json = try? JSONSerialization.jsonObject(with: payloadData) as? [String: Any] else { return nil }
            let isFinal = seq < 0 || msgFlags == flagNegSequence
            var text = ""
            if let result = json["result"] as? [String: Any] {
                text = result["text"] as? String ?? ""
                if text.isEmpty, let utts = result["utterances"] as? [[String: Any]] {
                    text = utts.compactMap { $0["text"] as? String }.joined()
                }
            }
            return ParsedResponse(kind: .result, sequence: seq, text: text, isFinal: isFinal)

        default:
            return nil
        }
    }
}

File: src-android/app/src/main/java/com/locke/murmur/VolcengineProtocol.kt
package com.locke.murmur

import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import java.util.zip.GZIPInputStream
import java.util.zip.GZIPOutputStream

object VolcengineProtocol {

    const val ENDPOINT = "wss://openspeech.bytedance.com/api/v3/sauc/bigmodel"

    // Message types (high nibble of byte 1)
    private const val MSG_FULL_CLIENT_REQUEST:  Byte = 0x01
    private const val MSG_AUDIO_ONLY_REQUEST:   Byte = 0x02
    private const val MSG_FULL_SERVER_RESPONSE: Byte = 0x09
    private const val MSG_SERVER_ACK:           Byte = 0x0B
    private const val MSG_SERVER_ERROR:         Byte = 0x0F

    // Sequence flags (low nibble of byte 1)
    private const val FLAG_POS_SEQUENCE: Byte = 0x01
    private const val FLAG_NEG_SEQUENCE: Byte = 0x03

    // Serialization (high nibble of byte 2)
    private const val SERIAL_JSON: Byte = 0x01

    // Compression (low nibble of byte 2)
    private const val COMPRESS_NONE: Byte = 0x00
    private const val COMPRESS_GZIP: Byte = 0x01

    private fun buildHeader(msgType: Byte, msgFlags: Byte, serial: Byte, compress: Byte): ByteArray {
        return byteArrayOf(
            ((0x01 shl 4) or 0x01).toByte(),        // version=1, headerSize=1
            ((msgType.toInt() shl 4) or msgFlags.toInt()).toByte(),
            ((serial.toInt() shl 4) or compress.toInt()).toByte(),
            0x00
        )
    }

    private fun int32ToBytes(value: Int): ByteArray =
        ByteBuffer.allocate(4).putInt(value).array()

    private fun bytesToInt32(bytes: ByteArray, offset: Int = 0): Int =
        ByteBuffer.wrap(bytes, offset, 4).int

    fun buildInitPacket(payload: JSONObject, sequence: Int): ByteArray {
        val jsonBytes = payload.toString().toByteArray(Charsets.UTF_8)
        val compressed = gzip(jsonBytes)

        val out = ByteArrayOutputStream()
        out.write(buildHeader(MSG_FULL_CLIENT_REQUEST, FLAG_POS_SEQUENCE, SERIAL_JSON, COMPRESS_GZIP))
        out.write(int32ToBytes(sequence))
        out.write(int32ToBytes(compressed.size))
        out.write(compressed)
        return out.toByteArray()
    }

    fun buildAudioPacket(audio: ByteArray, sequence: Int, isLast: Boolean): ByteArray {
        val flag = if (isLast) FLAG_NEG_SEQUENCE else FLAG_POS_SEQUENCE
        val seqValue = if (isLast) -sequence else sequence

        val out = ByteArrayOutputStream()
        out.write(buildHeader(MSG_AUDIO_ONLY_REQUEST, flag, SERIAL_JSON, COMPRESS_NONE))
        out.write(int32ToBytes(seqValue))
        out.write(int32ToBytes(audio.size))
        out.write(audio)
        return out.toByteArray()
    }

    data class ParsedResponse(
        val kind: Kind,
        val sequence: Int,
        val text: String? = null,
        val isFinal: Boolean = false,
        val errorMessage: String? = null
    ) {
        enum class Kind { ACK, RESULT, ERROR }
    }

    fun parseResponse(data: ByteArray): ParsedResponse? {
        if (data.size < 4) return null
        val msgType  = (data[1].toInt() ushr 4) and 0x0F
        val msgFlags =  data[1].toInt() and 0x0F
        val compress =  data[2].toInt() and 0x0F

        return when (msgType.toByte()) {
            MSG_SERVER_ERROR -> {
                if (data.size < 12) return null
                val msgSize = bytesToInt32(data, 8)
                if (msgSize < 0 || data.size < 12 + msgSize) return null
                val raw = data.copyOfRange(12, 12 + msgSize)
                val msg = if (compress == COMPRESS_GZIP.toInt()) ungzip(raw)?.toString(Charsets.UTF_8) ?: ""
                          else raw.toString(Charsets.UTF_8)
                ParsedResponse(ParsedResponse.Kind.ERROR, 0, errorMessage = msg)
            }
            MSG_SERVER_ACK -> {
                if (data.size < 8) return null
                val seq = bytesToInt32(data, 4)
                ParsedResponse(ParsedResponse.Kind.ACK, seq)
            }
            MSG_FULL_SERVER_RESPONSE -> {
                if (data.size < 12) return null
                val seq = bytesToInt32(data, 4)
                val payloadSize = bytesToInt32(data, 8)
                if (payloadSize < 0 || data.size < 12 + payloadSize) return null
                val rawPayload = data.copyOfRange(12, 12 + payloadSize)
                val payloadBytes = if (compress == COMPRESS_GZIP.toInt()) ungzip(rawPayload) ?: return null
                                   else rawPayload
                val json = runCatching { JSONObject(payloadBytes.toString(Charsets.UTF_8)) }.getOrNull() ?: return null
                val isFinal = seq < 0 || msgFlags == FLAG_NEG_SEQUENCE.toInt()
                var text = ""
                val result = json.optJSONObject("result")
                if (result != null) {
                    text = result.optString("text", "")
                    if (text.isEmpty()) {
                        val utts = result.optJSONArray("utterances")
                        if (utts != null) {
                            val sb = StringBuilder()
                            for (i in 0 until utts.length()) sb.append(utts.getJSONObject(i).optString("text", ""))
                            text = sb.toString()
                        }
                    }
                }
                ParsedResponse(ParsedResponse.Kind.RESULT, seq, text = text, isFinal = isFinal)
            }
            else -> null
        }
    }

    private fun gzip(data: ByteArray): ByteArray {
        val bos = ByteArrayOutputStream()
        GZIPOutputStream(bos).use { it.write(data) }
        return bos.toByteArray()
    }

    private fun ungzip(data: ByteArray): ByteArray? = runCatching {
        GZIPInputStream(data.inputStream()).use { it.readBytes() }
    }.getOrNull()
}
